### PR TITLE
watch internal TS improvement

### LINKED ISSUE #1371: Type definition for watch method should support returning non-string values
**Is your feature request related to a problem? Please describe.**
Current type definition for watch is:

```ts
watch<T extends FieldName<FormValues>>(field: T & string, defaultValue?: string): FormValues[T]
```

However limiting `defaultValue` to just `string` results in false positive ts errors. For example if working with a checkbox, I would want to do:

```ts
const isEnabled = watch("myCheckbox", true)
// results in ts error because true is not a string
```

**Describe the solution you'd like**
I propose changing this to:

```ts
watch<T extends keyof FormValues>(field: T, defaultValue?: FormValues[T]): FormValues[T]
```

Reasons:
* Allows us to pass non-string values as `defaultValue` without getting ts errors.
* Ensures greater type safety since `field` will need to be an existing key in `FormValues`.
* Ensures that `defaultValue` is a valid type based on the given `field`.
* Falls back gracefully to looser types if `FormValues` is not explicitly defined.

### PR DESCRIPTION
**Pull request by kotarella1110**
```ts
type FormValues = {
  key1: string;
  key2: number;
  key3: {
    key1: number;
    key2: boolean;
  };
};

const { watch } = useForm<FormValues>;

watch();
// function watch(): FormValues
watch({ nest: true })
// function watch(option: { nest: boolean; }): FormValues
watch('key1')
// function watch<"key1">(field: "key1", defaultValue?: string | undefined): string
watch('key1', 'test')
// function watch<"key1">(field: "key1", defaultValue?: string | undefined): string
watch('key1', true)
// ❌: type error
watch('key3.key1')
// function watch<unknown>(field: string, defaultValue?: unknown): unknown
watch('key3.key1', 1);
// function watch<1>(field: string, defaultValue?: 1 | undefined): number
watch('key3.key1', 'test');
// function watch<"key3.key1", "test">(field: "key3.key1", defaultValue?: string | undefined): string
watch('key3.key2', true);
// function watch<true>(field: string, defaultValue?: true | undefined): boolean
watch(['key1', 'key2'])
// function watch<"key1" | "key2">(fields: ("key1" | "key2")[], defaultValues?: DeepPartial<Pick<FormValues, "key1" | "key2">> | undefined): Pick<FormValues, "key1" | "key2">
watch(['key1', 'key2'], { key1: 'test' })
// function watch<"key1" | "key2">(fields: ("key1" | "key2")[], defaultValues?: DeepPartial<Pick<FormValues, "key1" | "key2">> | undefined): Pick<FormValues, "key1" | "key2">
watch(['key1', 'key2'], { key1: 'test', key2: true })
// ❌: type error
watch(['key1', 'key3.key1'], { key1: 'string' })
// function watch(fields: string[], defaultValues?: DeepPartial<FormValues> | undefined): DeepPartial<FormValues>
watch(['key1', 'key3.key1'], { test: 'string' })
// ❌: type error
```

[TypeScript Playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBAYg9gJwLYDUCGAbArhAzlAXigG8AoKKAawhAEYAuKXYBASwDsBzAbnKpoBMjdliQAjCAl4VqIAMyMyFGTQZQR4ydOWyhUMXDgYIadtoC+vS6VIQAHmETAooSFAAKbJK2CsAbtBEzGxcUAA++obGpuHqohIIsbgg4kaxIhgYsVjsACYQAGYcELm8tg5OLuDQADI+kpgAquyscOwAPKQAkAAqUPbAEHn4nqzevgEANN2N-XaDwx5ePv4QpAB8hFB9EQAUswBkJAD6APzCEAEI5gCUZa619QiYPXCj46vtfQNDufimIE2RG+81++GCHE4fFOTBYkL4jBBCz+cU0CGhqISCO2c2R+AMRhM7AxBOixIoiPu1SgABEIBAwO40AhfJgvkCSHwANruKAcfggOAFbYAXXO2x5ItxYKgAEEEM8QO0OAVJFBGus+BQYfLFe06QymSzWGyNZrlFBEZLpYsAEomXJtDAgXVoJUqtUarVQGH2tCO9jO11Kg2M5msjDtM3eq3uKU-RbEKBc2SMCFcEWMHKUdhwADu7Cg5m9MNDRojX0l5uUsZFVjK+QAxhhmdACjlG742lA82hgI2ABa7G6MeDIdDYPC8JsthBtjtdwu9-tDuBgReKdR4YCMUlEosj2CIVCYHC4acQZutqDt9id1pLvuDr42lF1QbPDDNB-tWRCo-jqeeCTLCIScOsIGzAmKLZrmBbrLsfBFBAGC5Ii0wUPkBRoFgGDABOODikiMp-sKY4npO+AwuRBF4FyPRSow74NBgrzvCsARRpqh7EYspEARRZ4+gJtG4PRjFQMxn5scsEwQFxF5XnON4Lg+PZPgOL7Qfg-E0UBuAIUhrAoX8VoihhUBYTheGieKZbhiakbuKwjaUO0emUSBPTrBBpCHs5rnuceoleesimzvOd6LupK6IRQyGobgaZwlwXLmXwVm4fh+l2fSYbGmyHlnr5h72QVkZFXgYU2Mug7DrwAD0DUqVFam1UOh6VbgpDtbsSbsNujAsDgB6kE1LX3t2vVrhuJBbswu5RES3AHqOwX6T1Gm7AA5LItDbTcY3Nbek2Piu7QAER7RdCEJWhUBXaoF0gZlNlAeK6acNkeSFMUuSHp9m2xbtqjbSB22DMwB1HRN0XtZd123cZqGMI9dDPZZhTWdlk4fSlX0RDkWF-QD+NA3VIN0GDLgIDgh3jYAMuRDdSkgIIg5NDpTcgAHR7dD40nXDGntLB+bsEjJnJWBL1Y1ltHiqLBaHorxK9VzvOgyBtB3DDgttcLtASyjoGQjL2Fy+9jC0N9xMDf9wjxJIHM7bIPN8+DkPAAdjXHapU3C2jbtPSBF2ezdux3ajrsa+jZvY-LUuQjbv126TYHO+rujU8NEA6wLftnc+OdG-dn1xxbuNDbT0CEz9RSp4thKmM7XKU-t4NZyK9O+61-vnWjtAXbEaMCOHd1JVAuwD0PEQjxdNxpeXb2UblhoOWyAVuV1IfXcPug3Zste2yU-kuVv62eQ9u+z-vmq9a37tQJTAjbeZc17YwEPbtto3573heaWnnvQQY9kamUnkAm+ICF5v1ejjM8q98oVk3kFQCl9IFXxAT5ZO9cT6MBQdvTB6NgEgFHnfLaD9NZP07iBJMH8n6e2proKuI1bgwyZlUNwrN2b3zbtTTOoM350NUJ-T6P82F-1OjFOq49E6pVgbLZeCDGBlQrF1Q+UAiYpzwbSPK5ZHKoMElVFufCO40CDlTIRLhBpPzEb-ZqHCHj9AVIgIAA)

close #1371

### CODE REVIEW

File: src/types.ts [hunk 1]
```diff
@@ -2,6 +2,14 @@ import * as React from 'react';
 
 export type Primitive = string | boolean | number | symbol | null | undefined;
 
+export type LiteralToPrimitive<T extends any> = T extends string
```
Comment (kotarella1110) [inline src/types.ts:5]: @bluebill1049 Is type alias naming "`LiteralToPremitive`" OK?

**Comment by bluebill1049** [comment]
WOW...

File: src/types.ts (hunk 1 above)
Comment (bluebill1049) [reply src/types.ts:5]: Sounds OK ✌🏻

File: src/contextTypes.ts [hunk 2]
```diff
@@ -50,21 +50,20 @@ export type FormContextValues<FormValues extends FieldValues = FieldValues> = {
...
-  watch<T extends FieldName<FormValues>>(
-    field: T & string,
-    defaultValue?: string,
-  ): FormValues[T];
+  watch<T extends string, U extends unknown>(
+    field: T,
+    defaultValue?: T extends keyof FormValues
+      ? FormValues[T]
+      : LiteralToPrimitive<U>,
+  ): T extends keyof FormValues ? FormValues[T] : LiteralToPrimitive<U>;
```
Comment (bluebill1049) [inline src/contextTypes.ts:58]: wow... super smart.

**Review by bluebill1049 — APPROVED** [review]

**Comment by bluebill1049** [comment]
> ```ts
> type FormValues = {
>   key1: string;
>   key2: number;
>   key3: {
>     key1: number;
>     key2: boolean;
>   };
> };
> 
> const { watch } = useForm<FormValues>;
> 
> watch();
> // function watch(): FormValues
> watch({ nest: true })
> // function watch(option: { nest: boolean; }): FormValues
> watch('key1')
> // function watch<"key1">(field: "key1", defaultValue?: string | undefined): string
> watch('key1', 'test')
> // function watch<"key1">(field: "key1", defaultValue?: string | undefined): string
> watch('key1', true)
> // ❌: type error
> watch('key3.key1')
> // function watch<unknown>(field: string, defaultValue?: unknown): unknown
> watch('key3.key1', 1);
> // function watch<1>(field: string, defaultValue?: 1 | undefined): number
> watch('key3.key1', 'test');
> // function watch<"key3.key1", "test">(field: "key3.key1", defaultValue?: string | undefined): string
> watch('key3.key2', true);
> // function watch<true>(field: string, defaultValue?: true | undefined): boolean
> watch(['key1', 'key2'])
> // function watch<"key1" | "key2">(fields: ("key1" | "key2")[], defaultValues?: DeepPartial<Pick<FormValues, "key1" | "key2">> | undefined): Pick<FormValues, "key1" | "key2">
> watch(['key1', 'key2'], { key1: 'test' })
> // function watch<"key1" | "key2">(fields: ("key1" | "key2")[], defaultValues?: DeepPartial<Pick<FormValues, "key1" | "key2">> | undefined): Pick<FormValues, "key1" | "key2">
> watch(['key1', 'key2'], { key1: 'test', key2: true })
> // ❌: type error
> watch(['key1', 'key3.key1'], { key1: 'string' })
> // function watch(fields: string[], defaultValues?: DeepPartial<FormValues> | undefined): DeepPartial<FormValues>
> watch(['key1', 'key3.key1'], { test: 'string' })
> // ❌: type error
> ```
> 
> [TypeScript Playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBAYg9gJwLYDUCGAbArhAzlAXigG8AoKKAawhAEYAuKXYBASwDsBzAbnKpoBMjdliQAjCAl4VqIAMyMyFGTQZQR4ydOWyhUMXDgYIadtoC+vS6VIQAHmETAooSFAAKbJK2CsAbtBEzGxcUAA++obGpuHqohIIsbgg4kaxIhgYsVjsACYQAGYcELm8tg5OLuDQADI+kpgAquyscOwAPKQAkAAqUPbAEHn4nqzevgEANN2N-XaDwx5ePv4QpAB8hFB9EQAUswBkJAD6APzCEAEI5gCUZa619QiYPXCj46vtfQNDufimIE2RG+81++GCHE4fFOTBYkL4jBBCz+cU0CGhqISCO2c2R+AMRhM7AxBOixIoiPu1SgABEIBAwO40AhfJgvkCSHwANruKAcfggOAFbYAXXO2x5ItxYKgAEEEM8QO0OAVJFBGus+BQYfLFe06QymSzWGyNZrlFBEZLpYsAEomXJtDAgXVoJUqtUarVQGH2tCO9jO11Kg2M5msjDtM3eq3uKU-RbEKBc2SMCFcEWMHKUdhwADu7Cg5m9MNDRojX0l5uUsZFVjK+QAxhhmdACjlG742lA82hgI2ABa7G6MeDIdDYPC8JsthBtjtdwu9-tDuBgReKdR4YCMUlEosj2CIVCYHC4acQZutqDt9id1pLvuDr42lF1QbPDDNB-tWRCo-jqeeCTLCIScOsIGzAmKLZrmBbrLsfBFBAGC5Ii0wUPkBRoFgGDABOODikiMp-sKY4npO+AwuRBF4FyPRSow74NBgrzvCsARRpqh7EYspEARRZ4+gJtG4PRjFQMxn5scsEwQFxF5XnON4Lg+PZPgOL7Qfg-E0UBuAIUhrAoX8VoihhUBYTheGieKZbhiakbuKwjaUO0emUSBPTrBBpCHs5rnuceoleesimzvOd6LupK6IRQyGobgaZwlwXLmXwVm4fh+l2fSYbGmyHlnr5h72QVkZFXgYU2Mug7DrwAD0DUqVFam1UOh6VbgpDtbsSbsNujAsDgB6kE1LX3t2vVrhuJBbswu5RES3AHqOwX6T1Gm7AA5LItDbTcY3Nbek2Piu7QAER7RdCEJWhUBXaoF0gZlNlAeK6acNkeSFMUuSHp9m2xbtqjbSB22DMwB1HRN0XtZd123cZqGMI9dDPZZhTWdlk4fSlX0RDkWF-QD+NA3VIN0GDLgIDgh3jYAMuRDdSkgIIg5NDpTcgAHR7dD40nXDGntLB+bsEjJnJWBL1Y1ltHiqLBaHorxK9VzvOgyBtB3DDgttcLtASyjoGQjL2Fy+9jC0N9xMDf9wjxJIHM7bIPN8+DkPAAdjXHapU3C2jbtPSBF2ezdux3ajrsa+jZvY-LUuQjbv126TYHO+rujU8NEA6wLftnc+OdG-dn1xxbuNDbT0CEz9RSp4thKmM7XKU-t4NZyK9O+61-vnWjtAXbEaMCOHd1JVAuwD0PEQjxdNxpeXb2UblhoOWyAVuV1IfXcPug3Zste2yU-kuVv62eQ9u+z-vmq9a37tQJTAjbeZc17YwEPbtto3573heaWnnvQQY9kamUnkAm+ICF5v1ejjM8q98oVk3kFQCl9IFXxAT5ZO9cT6MBQdvTB6NgEgFHnfLaD9NZP07iBJMH8n6e2proKuI1bgwyZlUNwrN2b3zbtTTOoM350NUJ-T6P82F-1OjFOq49E6pVgbLZeCDGBlQrF1Q+UAiYpzwbSPK5ZHKoMElVFufCO40CDlTIRLhBpPzEb-ZqHCHj9AVIgIAA)
> 
> close #1371

Awesome work!!! Let's merge this PR! ❤️ 

I think we should add some TS tests in this lib. Maybe let you drive on this? This is a task on our project board.

**Comment by bluebill1049** [comment]
Screen Shot 2020-04-13 at 9 03 06 am

In fact, maybe we can add these into the unit tests as well. It will stop the build as well. Your thoughts?

**Comment by kotarella1110** [comment]
@bluebill1049 sounds good! I will try type test 💯

**Comment by bluebill1049** [comment]
actually @kotarella1110 can we do the same with `getValues` TS support with single field lookup?

**Comment by kotarella1110** [comment]
@bluebill1049 Yeah! I will try to confirm about whether we can do it 👍

**Comment by bluebill1049** [comment]
Dancing!

**Comment by jsbrain** [comment]
Hi, I don't get it... I'm still getting the same result as before:

```ts
export type FormData = {
  form: {
    title: string
    text: string
  }
}

const methods = useForm<LeadListFormData, FormValidationContext>({
    validationContext: { entity: 'Link' },
    validateCriteriaMode: 'all',
    submitFocusError: true,
})

const previewFormTitle = watch('form.title')
//  => unknown
const previewFormText = watch('form.text')
// => unknown

```

What am I missing?

**Comment by kotarella1110** [comment]
@jsbrain `unknown` is correct because TypeScript cannot infer the return value of the nested object.
You can define the return type by passing defaultValue or generics to `watch`.

```ts
const previewFormTitle = watch('form.title')
//  => unknown
const previewFormText = watch('form.text')
// => unknown
const previewFormTitle = watch('form.title', 'default title')
//  => string
const previewFormText = watch('form.text', 'default text')
//  => string
const previewFormText = watch<string, FormData['form']['title']>('form.title')
//  => string
const previewFormText = watch<string, FormData['form']['text']>('form.text')
//  => string
```

If it is a flat object, you don't need to pass defaultValue or generics.
it will be inferred correctly.

```ts
export type FormData = {
  title: string;
  text: string;
}

const methods = useForm<FormData, FormValidationContext>({
    validationContext: { entity: 'Link' },
    validateCriteriaMode: 'all',
    submitFocusError: true,
})

const previewFormTitle = watch('title')
//  => string
const previewFormText = watch('.text')
//  => string
```

**Comment by jsbrain** [comment]
Oh perfect. I was missing the generics. Thank you!

**Comment by bluebill1049** [comment]
@jsbrain we going to improve the Ts in the next major version and improve the docs as well. Hopefully, we can have enough content to create a dedicated page for TS support.

**Comment by jsbrain** [comment]
@bluebill1049  sounds good! The generics version isn't documented yet or did I actually not find it in the docs?

**Comment by bluebill1049** [comment]
@jsbrain it's missing in the doc, i will get that updated tonight :)

**Comment by jsbrain** [comment]
Cool, love it. One more question:

This is what I'm actually doing, I got an input in one component with a default value and another component which shows a `preview` of the values.

Simplified, it looks like this:

```tsx
type PreviewFormData = {
  title: string
  description: string
  form: {
    title: string
    text: string
  }
}

const PreviewForm: React.FC<{}> = () => {
  const defaultValues = {
    form: {
      title: 'Preview title',
      text: 'Preview text',
    },
  }

  const { register, watch } = useForm<PreviewFormData>({
    validateCriteriaMode: 'all',
    submitFocusError: true,
  })

  // Form Field Validators
  const formValidators = {
    form: {
      title: register({
        // required: 'This is required.',
        maxLength: {
          value: 80,
          message: 'Must not be longer than 80 characters',
        },
        minLength: {
          value: 3,
          message: 'Title must be 3 characters minimum',
        },
      }),
      text: register({
        maxLength: {
          value: 220,
          message: 'Must not be longer than 220 characters',
        },
        minLength: {
          value: 3,
          message: 'If set, must be 3 characters minimum',
        },
      }),
    },
  }

  // Form Preview
  const previewFormTitle = watch('form.title', defaultValues.form.title)
  const previewFormText = watch('form.text', defaultValues.form.text)

  return (
    <div>
      {/* Component with input showing default value */}
      <div>
        <input
          type="text"
          ref={formValidators.form.title}
          defaultValue={defaultValues.form.title}
          name="form.title"
        />
        <input
          type="text"
          ref={formValidators.form.text}
          defaultValue={defaultValues.form.text}
          name="form.text"
        />
      </div>

      {/* Other component showing preview */}
      <div>
        <div>{previewFormTitle}</div>
        <div>{previewFormText}</div>
      </div>
    </div>
  )
}
```

So I noticed for this to actually work I have to use the defaultValue on both, the component and on the watch. This feels somehow a little bit counter-intuitive for me but is this the expected behavior or is there a better way to do this?

**Comment by bluebill1049** [comment]
no, you shouldn't need both `defaultValues` and `defaultValue`. it should be either.

**Comment by bluebill1049** [comment]
if you defined defaultValues, watch will take `defautValue` over as well.

**Comment by bluebill1049** [comment]
https://codesandbox.io/s/competent-cookies-kfcjm?file=/src/App.js

**Comment by kotarella1110** [comment]
@bluebill1049 we need to `defautValue` when it is nested object.

**Comment by kotarella1110** [comment]
@jsbrain In such cases it is better to use generics.

```ts
  // Form Preview
  const previewFormTitle = watch<string, PreviewFormData['form']['title']>('form.title', defaultValues.form.title)
  const previewFormText = watch<string, PreviewFormData['form']['text']>('form.text', defaultValues.form.text)
```

**Comment by jsbrain** [comment]
@bluebill1049 I actually tried this 2 days before but it did not work for me on my component. It is pretty complicated tho (about 700 loc). Tried to reproduce the issue but found the fault in my code. In a useEffect hook I did `reset({})` at some point for testing stuff and totally forgot about this... 🤦🏻‍♂️

Thank for making me find this error in my bloated component haha

**Comment by bluebill1049** [comment]
hahaha. no worries at all @jsbrain we all make silly mistakes. next time if you encounter an issue with RHF, feel free to send a question at our Specturm channel, some of us can help you out :)

**Comment by jsbrain** [comment]
Good point! Will do :D

**Pull request merged**